File: 2025/06/python/solution.py
from functools import reduce
from itertools import zip_longest
import sys
import operator

OPERATIONS = {"*": operator.mul, "+": operator.add}

# Token grid -> TODO: Add to utility templates, port to other language templates if possible


def read_input_part1():
    with open('./input', 'r') as file:
        for line in file:
            yield line.split()

# Character grid -> TODO: Add to utility templates, this is a nice approach!


def read_input_part2():
    with open('./input', 'r') as file:
        for line in file:
            yield list(line.rstrip("\n") + " ")


def parse(digits):
    s = ''.join(digits).strip()
    return None if not s else int(s)


def part_one(rows):
    total = 0
    for col in zip_longest(*rows):
        *nums, op_symbol = col
        op = OPERATIONS[op_symbol]
        digits = (int(n) for n in nums if n is not None)
        total += reduce(op, digits)
    return total


def part_two(rows):
    total = 0
    op = None
    digits = []

    for col in zip_longest(*rows, fillvalue=' '):
        data = list(col)
        key = data.pop()
        if key in OPERATIONS:
            op = OPERATIONS[key]

        value = parse(data)
        if value is not None:
            digits.append(value)
        else:
            if digits and op is not None:
                result = reduce(op, digits)
                total += result
                digits = []

    if digits and op is not None:
        total += reduce(op, digits)

    return total


def main() -> None:
    arg = sys.argv[1] if len(sys.argv) > 1 else "both"

    if arg == '1':
        print(part_one(read_input_part1()))
    elif arg == '2':
        print(part_two(read_input_part2()))
    else:
        print(part_one(read_input_part1()))
        print(part_two(read_input_part2()))


if __name__ == '__main__':
    main()

File: 2025/06/elixir/solution.ex
defmodule Solution do
  @input File.read!("input")

  @operations %{
    "*" => &Kernel.*/2,
    "+" => &Kernel.+/2
  }

  # Parsing
  # Token grid
  defp read_input_part1(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(&String.split/1)
  end

  # Char grid
  defp read_input_part2(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      (line <> " ") |> String.graphemes()
    end)
  end

  # Parse a list of single-char strings into an int
  defp parse_digits(chars) do
    s = chars |> Enum.join() |> String.trim()
    if s == "", do: nil, else: String.to_integer(s)
  end

  defp columns(rows, fill \\ nil) do
    max_cols =
      rows
      |> Enum.map(&length/1)
      |> Enum.max(fn -> 0 end)

    for idx <- 0..(max_cols - 1) do
      Enum.map(rows, &Enum.at(&1, idx, fill))
    end
  end

  # Solutions
  def part1(input \\ @input) do
    rows = read_input_part1(input)

    rows
    |> columns()
    |> Enum.reduce(0, fn col, acc ->
      {nums, [op_symbol]} = Enum.split(col, -1)
      op = Map.fetch!(@operations, op_symbol)

      digits =
        nums
        |> Enum.reject(&is_nil/1)
        |> Enum.map(&String.to_integer/1)

      acc + Enum.reduce(digits, op)
    end)
  end

  def part2(input \\ @input) do
    rows = read_input_part2(input)

    {total, op, digits} =
      rows
      |> columns(" ")
      |> Enum.reduce({0, nil, []}, fn col, {total, op, digits} ->
        {data, [key]} = Enum.split(col, -1)

        op =
          case Map.get(@operations, key) do
            nil -> op
            fun -> fun
          end

        value = parse_digits(data)

        cond do
          value != nil ->
            {total, op, [value | digits]}

          digits != [] and op != nil ->
            result = digits |> Enum.reverse() |> Enum.reduce(op)
            {total + result, op, []}

          true ->
            {total, op, digits}
        end
      end)

    if digits != [] and op != nil do
      total + (digits |> Enum.reverse() |> Enum.reduce(op))
    else
      total
    end
  end
end

IO.puts(Solution.part1())
IO.puts(Solution.part2())
